File: sdk/php/src/Retry/RetryPolicy.php
<?php

declare(strict_types=1);

namespace Velocity\SDK\Retry;

use InvalidArgumentException;

/**
 * Retry policy with exponential backoff and jitter.
 *
 * Usage:
 *     $policy = RetryPolicy::builder()
 *         ->maxAttempts(5)
 *         ->initialIntervalMs(100)
 *         ->backoffCoefficient(2.0)
 *         ->maxIntervalMs(10000)
 *         ->jitter(true)
 *         ->build();
 *
 *     $result = $policy->execute(function() {
 *         return fetchRemoteData();
 *     });
 */
class RetryPolicy
{
    private int $maxAttempts;
    private int $initialIntervalMs;
    private float $backoffCoefficient;
    private int $maxIntervalMs;
    private bool $jitter;
    /** @var array<class-string<\Throwable>> */
    private array $retryableExceptions;

    /**
     * @param array<class-string<\Throwable>> $retryableExceptions
     */
    public function __construct(
        int $maxAttempts = 3,
        int $initialIntervalMs = 100,
        float $backoffCoefficient = 2.0,
        int $maxIntervalMs = 60000,
        bool $jitter = true,
        array $retryableExceptions = []
    ) {
        $this->maxAttempts = $maxAttempts;
        $this->initialIntervalMs = $initialIntervalMs;
        $this->backoffCoefficient = $backoffCoefficient;
        $this->maxIntervalMs = $maxIntervalMs;
        $this->jitter = $jitter;
        $this->retryableExceptions = $retryableExceptions;
        $this->validate();
    }

    /** Create a default retry policy. */
    public static function defaults(): self
    {
        return new self();
    }

    /** Create a new builder. */
    public static function builder(): RetryPolicyBuilder
    {
        return new RetryPolicyBuilder();
    }

    /**
     * Execute a callable with retry logic.
     *
     * @template T
     * @param callable(): T $callable
     * @return T
     * @throws \Throwable
     */
    public function execute(callable $callable): mixed
    {
        $this->validate();

        $lastException = null;

        for ($attempt = 0; $attempt < $this->maxAttempts; $attempt++) {
            try {
                return $callable();
            } catch (\Throwable $e) {
                $lastException = $e;

                if (!$this->isRetryable($e)) {
                    throw $e;
                }

                if ($attempt < $this->maxAttempts - 1) {
                    $backoff = $this->calculateBackoff($attempt);
                    usleep((int) ($backoff * 1000));
                }
            }
        }

        throw $lastException;
    }

    /** Calculate backoff duration in milliseconds for a given attempt. */
    public function calculateBackoff(int $attempt): float
    {
        $interval = $this->initialIntervalMs * ($this->backoffCoefficient ** $attempt);
        $interval = min($interval, $this->maxIntervalMs);

        if ($this->jitter) {
            $interval = mt_rand(0, (int) $interval);
        }

        return $interval;
    }

    /** Check if an exception is retryable. */
    public function isRetryable(\Throwable $e): bool
    {
        if (empty($this->retryableExceptions)) {
            return true; // retry all by default
        }

        foreach ($this->retryableExceptions as $class) {
            if ($e instanceof $class) {
                return true;
            }
        }

        return false;
    }

    private function validate(): void
    {
        if ($this->maxAttempts < 1) {
            throw new InvalidArgumentException('maxAttempts must be >= 1');
        }
        if ($this->initialIntervalMs <= 0) {
            throw new InvalidArgumentException('initialIntervalMs must be > 0');
        }
        if ($this->backoffCoefficient < 1.0) {
            throw new InvalidArgumentException('backoffCoefficient must be >= 1.0');
        }
        if ($this->maxIntervalMs < $this->initialIntervalMs) {
            throw new InvalidArgumentException('maxIntervalMs must be >= initialIntervalMs');
        }
    }

    // ─── Getters ───────────────────────────────────────────────────────────────

    public function getMaxAttempts(): int { return $this->maxAttempts; }
    public function getInitialIntervalMs(): int { return $this->initialIntervalMs; }
    public function getBackoffCoefficient(): float { return $this->backoffCoefficient; }
    public function getMaxIntervalMs(): int { return $this->maxIntervalMs; }
    public function isJitterEnabled(): bool { return $this->jitter; }
}

/**
 * Builder for RetryPolicy.
 */
class RetryPolicyBuilder
{
    private int $maxAttempts = 3;
    private int $initialIntervalMs = 100;
    private float $backoffCoefficient = 2.0;
    private int $maxIntervalMs = 60000;
    private bool $jitter = true;
    /** @var array<class-string<\Throwable>> */
    private array $retryableExceptions = [];

    public function maxAttempts(int $maxAttempts): self
    {
        $this->maxAttempts = $maxAttempts;
        return $this;
    }

    public function initialIntervalMs(int $ms): self
    {
        $this->initialIntervalMs = $ms;
        return $this;
    }

    public function backoffCoefficient(float $coeff): self
    {
        $this->backoffCoefficient = $coeff;
        return $this;
    }

    public function maxIntervalMs(int $ms): self
    {
        $this->maxIntervalMs = $ms;
        return $this;
    }

    public function jitter(bool $enabled): self
    {
        $this->jitter = $enabled;
        return $this;
    }

    /**
     * @param class-string<\Throwable> $exceptionClass
     */
    public function addRetryableException(string $exceptionClass): self
    {
        $this->retryableExceptions[] = $exceptionClass;
        return $this;
    }

    public function build(): RetryPolicy
    {
        return new RetryPolicy(
            $this->maxAttempts,
            $this->initialIntervalMs,
            $this->backoffCoefficient,
            $this->maxIntervalMs,
            $this->jitter,
            $this->retryableExceptions,
        );
    }
}

File: sdk/java/src/main/java/io/velocity/sdk/retry/RetryPolicy.java
package io.velocity.sdk.retry;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

/**
 * Retry policy with exponential backoff and jitter.
 *
 * <p>Usage:
 * <pre>{@code
 * RetryPolicy policy = RetryPolicy.builder()
 *     .maxAttempts(5)
 *     .initialIntervalMs(100)
 *     .backoffCoefficient(2.0)
 *     .maxIntervalMs(10_000)
 *     .jitter(true)
 *     .addRetryableException(IOException.class)
 *     .build();
 *
 * String result = policy.execute(() -> fetchRemoteData());
 * }</pre>
 */
public class RetryPolicy {

    private final int maxAttempts;
    private final long initialIntervalMs;
    private final double backoffCoefficient;
    private final long maxIntervalMs;
    private final boolean jitter;
    private final Set<Class<? extends Exception>> retryableExceptions;

    private RetryPolicy(Builder builder) {
        this.maxAttempts = builder.maxAttempts;
        this.initialIntervalMs = builder.initialIntervalMs;
        this.backoffCoefficient = builder.backoffCoefficient;
        this.maxIntervalMs = builder.maxIntervalMs;
        this.jitter = builder.jitter;
        this.retryableExceptions = builder.retryableExceptions;
        validate();
    }

    /** Create a new builder for RetryPolicy. */
    public static Builder builder() {
        return new Builder();
    }

    /** Create a default retry policy (3 attempts, 100ms initial, 2x backoff). */
    public static RetryPolicy defaults() {
        return builder().build();
    }

    /**
     * Execute a callable with retry logic.
     *
     * @param callable the operation to execute
     * @param <T>      the return type
     * @return the result of the callable
     * @throws Exception the last exception if all retries fail
     */
    public <T> T execute(Callable<T> callable) throws Exception {
        validate();
        Exception lastException = null;

        for (int attempt = 0; attempt < maxAttempts; attempt++) {
            try {
                return callable.call();
            } catch (Exception e) {
                lastException = e;

                if (!isRetryable(e)) {
                    throw e;
                }

                if (attempt < maxAttempts - 1) {
                    long backoff = calculateBackoff(attempt);
                    try {
                        TimeUnit.MILLISECONDS.sleep(backoff);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw e;
                    }
                }
            }
        }

        throw lastException;
    }

    /**
     * Execute a runnable with retry logic (void return).
     *
     * @param runnable the operation to execute
     * @throws Exception the last exception if all retries fail
     */
    public void execute(Runnable runnable) throws Exception {
        execute(() -> {
            runnable.run();
            return null;
        });
    }

    /** Calculate backoff duration for a given attempt. */
    public long calculateBackoff(int attempt) {
        double interval = initialIntervalMs * Math.pow(backoffCoefficient, attempt);
        interval = Math.min(interval, maxIntervalMs);

        if (jitter) {
            interval = ThreadLocalRandom.current().nextDouble() * interval;
        }

        return (long) interval;
    }

    /** Check if an exception is retryable. */
    public boolean isRetryable(Exception e) {
        if (retryableExceptions.isEmpty()) {
            return true; // retry all by default
        }
        for (Class<? extends Exception> cls : retryableExceptions) {
            if (cls.isInstance(e)) {
                return true;
            }
        }
        return false;
    }

    private void validate() {
        if (maxAttempts < 1) throw new IllegalArgumentException("maxAttempts must be >= 1");
        if (initialIntervalMs <= 0) throw new IllegalArgumentException("initialIntervalMs must be > 0");
        if (backoffCoefficient < 1.0) throw new IllegalArgumentException("backoffCoefficient must be >= 1.0");
        if (maxIntervalMs < initialIntervalMs) throw new IllegalArgumentException("maxIntervalMs must be >= initialIntervalMs");
    }

    // ─── Getters ───────────────────────────────────────────────────────────────

    public int getMaxAttempts() { return maxAttempts; }
    public long getInitialIntervalMs() { return initialIntervalMs; }
    public double getBackoffCoefficient() { return backoffCoefficient; }
    public long getMaxIntervalMs() { return maxIntervalMs; }
    public boolean isJitter() { return jitter; }

    // ─── Builder ───────────────────────────────────────────────────────────────

    public static class Builder {
        private int maxAttempts = 3;
        private long initialIntervalMs = 100;
        private double backoffCoefficient = 2.0;
        private long maxIntervalMs = 60_000;
        private boolean jitter = true;
        private final Set<Class<? extends Exception>> retryableExceptions = new HashSet<>();

        public Builder maxAttempts(int maxAttempts) {
            this.maxAttempts = maxAttempts;
            return this;
        }

        public Builder initialIntervalMs(long initialIntervalMs) {
            this.initialIntervalMs = initialIntervalMs;
            return this;
        }

        public Builder backoffCoefficient(double backoffCoefficient) {
            this.backoffCoefficient = backoffCoefficient;
            return this;
        }

        public Builder maxIntervalMs(long maxIntervalMs) {
            this.maxIntervalMs = maxIntervalMs;
            return this;
        }

        public Builder jitter(boolean jitter) {
            this.jitter = jitter;
            return this;
        }

        public Builder addRetryableException(Class<? extends Exception> exceptionClass) {
            this.retryableExceptions.add(exceptionClass);
            return this;
        }

        public RetryPolicy build() {
            return new RetryPolicy(this);
        }
    }
}
